File: src/main/java/br/com/rfasioli/micrometermetrics/resource/mapper/PersonMapper.java
package br.com.rfasioli.micrometermetrics.resource.mapper;

import br.com.rfasioli.micrometermetrics.persistence.document.Person;
import br.com.rfasioli.micrometermetrics.resource.dto.PersonDTO;
import org.bson.types.ObjectId;
import org.mapstruct.Mapper;
import org.mapstruct.factory.Mappers;

@Mapper(componentModel="spring")
public interface PersonMapper {
  PersonMapper INSTANCE = Mappers.getMapper( PersonMapper.class );

  PersonDTO toPersonDTO(Person person);

  Person fromPersonDTO(PersonDTO dto);

  default String map(ObjectId value) {
    return value != null ? value.toHexString() : null;
  }

  default ObjectId map(String value) {
    return value != null && ObjectId.isValid(value) ? new ObjectId(value) : null;
  }
}

File: src/main/kotlin/br/com/rfasioli/micrometermetrics/resource/mapper/PersonMapper.kt
package br.com.rfasioli.micrometermetrics.resource.mapper;

import br.com.rfasioli.micrometermetrics.persistence.document.Person
import br.com.rfasioli.micrometermetrics.resource.dto.PersonDTO
import org.bson.types.ObjectId

fun Person.toPersonDTO() =
    PersonDTO(
        id = this.id?.toHexString(),
        firstName = this.firstName,
        sureName = this.sureName,
        lastName = this.lastName,
    )

fun PersonDTO.toPerson() =
    Person(
        id = if (this.id != null && ObjectId.isValid(this.id)) ObjectId(this.id) else null,
        firstName = this.firstName,
        sureName = this.sureName,
        lastName = this.lastName,
    )
